crates/lib3h/src/rrdht_util/r_value.rs

```diff
@@ -0,0 +1,335 @@
+}
+
+impl Default for RValuePeerRecordSet {
+    fn default() -> Self {
+        Self {
+            arc_of_included_peer_records: Arc::new(0.into(), 0),
+            peer_records: Vec::new(),
+        }
+    }
+}
+
+impl RValuePeerRecordSet {
+    /// the arc we are representing with this PeerRecordSet
+    pub fn arc_of_included_peer_records(mut self, arc_of_included_peer_records: Arc) -> Self {
+        self.arc_of_included_peer_records = arc_of_included_peer_records;
+        self
+    }
+
+    /// push a new PeerRecord into this record set
+    pub fn push_peer_record(mut self, peer_record: RValuePeerRecord) -> Self {
+        if !self
+            .arc_of_included_peer_records
+            .contains_location(peer_record.storage_arc.center())
+        {
+            panic!("peer record does not fit within arc_of_included_peer_records");
+        }
+        self.peer_records.push(peer_record);
+        self
+    }
+}
+
+/// Interpolate qualified redundancy in the network given our current
+/// knowledge of peers' claimed storage arcs.
+/// For a given storage arc (should match *this* agent's storage arc),
+/// if all the peers we know about are actually storing what they claim to be
+/// we can interpolate what the whole network "R" value is
+/// also taking into account our experience of those agents' reachability.
+pub fn interpolate_r_value_for_given_arc(peer_record_set: &RValuePeerRecordSet) -> f64 {
+    let mut running_avg_coverage = 0.0_f64;
+
+    for (i, record) in peer_record_set.peer_records.iter().enumerate() {
+        // what is the total coverage factor for this record?
+        let coverage =
+            (record.storage_arc.length() as f64 / ARC_LENGTH_MAX as f64) * record.uptime_0_to_1;
+
+        // mix it into our running average
+        running_avg_coverage = (running_avg_coverage * i as f64 + coverage) / (i as f64 + 1.0);
+    }
+
+    // given our known arc, estimate the total node count
+    let est_node_count = peer_record_set.peer_records.len() as f64
+        * (1.0
+            / (peer_record_set.arc_of_included_peer_records.length() as f64
+                / ARC_LENGTH_MAX as f64));
+
+    // get the estimated total coverage a.k.a r-value
+    running_avg_coverage * est_node_count
+}
+
+/// As an agent, we need to set our storage arc radius to something
+/// on the one hand, if we have enough storage and network bandwidth,
+/// we might want to set it as high as possible. On the other hand, we
+/// might want to conserve resources / power. If our network is healthy,
+/// we can scale back a bit to conserve, if it is in danger, we may want
+/// to scale up for safety. If it is anywhere in the middle, we probably
+/// want to maintain stability so we're not thrashing our resources.
+///
+/// Given current network conditions (within the sample slice given)
+/// this function will return a recommended storage arc radius.
+/// RADIUS not length. Please use `Arc::new_radius(0.into(), radius).length()`
```

Comment:
Using an arc radius (which could be a rather confusing concept in itself since it's not referring to the radius of the circle which the arc is on, but rather the 1/2 size of an arc that is symmetric around a center location) limits us to using symmetric storage arcs.

I want us to eventually be free from the assumption that a storage arc is symmetric around the address location of the peer. I sense it's a special case that we're committing to and there will be benefits if we can more easily explore the general case problem later. Are there some minor adjustments that could be made here that allow us to more easily transition later?

Reply:
See my above comment about minimizing gossip data.

I don't feel like I'm baking us into any corners here... we can and will need to alter this api to accept additional metrics for determining the best arc for given agents, so we shouldn't feel like this MVP is the end-all of the conversation.

two other things specifically about non-centered arcs though:
- if we have so few people on our network that the specific positioning of their arc relative to their center location makes any difference, then our network is unhealthy and everyone should hold full arcs anyways
- I think giving agents control of how their arcs are oriented may be too game-able... OH if i just stretch in this direction a bit, I can have a voice in that entries validation!